db/migrate/20190601004310_create_one_time_tasker_task_attempts.rb

```diff
@@ -0,0 +1,13 @@
+class CreateOneTimeTaskerTaskAttempts < ActiveRecord::Migration[5.2]
+  def change
+    create_table :one_time_tasker_task_attempts do |t|
+      t.string :task_name, null: false
+      t.string :task_source
+      t.timestamp :attempted_on, null: false
+      t.boolean :was_successful, null: false, default: false
+      t.string :notes, null: true
+
+      t.timestamps
+    end
+  end
+end
```

Comment:
Relative to my previous comments, I'm still uneasy about using the DB as our source of knowledge as to which tasks have been run and which have not.  The assumption here is that the DB accurately specifies which tasks have been run and which have not, on _this instance_ of the DB.

For production we're probably OK with that assumption (although we'd be in trouble if - for some reason - we failed to recover this table when restoring the DB after some disaster - but that's a low probability event).

The problem exists, however, for development and staging environments.  If I clone the project, run seeds and then run this, I could get all kinds of strange errors - at rake runtime, and maybe later with data that was altered by a one-time task that should not have been.   Or, I might run the task later when I've build up a lot of local data, and that data might be altered in ways not intended.

I am wondering if we should restrict this rake task to running _only in production_?

Reply:
In this app, in any environment, if you run a rake task that puts the data into a bad state, then your data is in a bad state.  
Protecting against putting data in a bad state is a _big_ task and not something we've ever really discussed or devoted time to.  (There are definitely apps where that is _critical_ and fundamental, but I don't think this is one of them.) 
For sure, `db:seed` is something that really should only ever been run once in production.  We have guards in there to not run in production really as a brute-force (and common) way to ensure that it is not run again. We use that as a simple, brute-force way to implement "run once."

So I think that is outside of the scope of this.  

If you feel strongly, then can we make that a separate PR so that we can contain the scope of this and try to get this PR wrapped up?

(As an aside: The purpose of a staging environment is to mirror the production environment as closely as possible so that any errors are caught in staging; they never get past staging to get into production.  (Plus, we don't have a staging environment.)  Maybe you mean to refer to "test"?)

Reply:
Re: the DB

Is your concern about using the DB more about using the application's db to keep track of the state of the data in the application?  IOW, is this more about keeping the information about the "meta state" of the data  outside of the application?

That is definitely something to consider.

The value of using a DB is that you get all of the capabilities and power of a DB (obv): the timestamps, the capabilities of db-management (dumps, schema control, etc etc)., ability to put the DB in separate volumes/disks/etc.

Alternatives would be to track things in a text file (e.g. YAML).  

Worth discussing, but would like to control the scope of this PR, so perhaps this could be done as a next step.

Reply:
My concern is using the DB for keeping track of the operations (rake tasks) that have been run against the data in that DB.  If the meta data is incomplete or missing, and a user runs the master run-once rake task, that inadvertent changes could occur to the DB.

For production, we're probably OK - since we control that DB and the running of rake tasks.

In development, we could run into the situation where the user has seeded the DB and also added much data during development, then runs the master rake task - which might change the dev DB in ways that are not visible but that create problems for the application and test logic.  That's why I think it might be good to guard against this rake tasks from running _outside_ of production.  Any specific run-once rake tasks that the user needs to run can be run individually, and with the user (presumably) aware of why the task should be run and what it will do to the dev DB.